iroha/src/kura.rs

```diff
@@ -180,126 +196,250 @@ impl Default for Mode {
     ///
     /// # Errors
     /// - Failed to create directory.
-    pub fn new(path: &Path) -> Result<BlockStore> {
-        if fs::read_dir(path).is_err() {
-            fs::create_dir_all(path).wrap_err("Failed to create Block Store directory.")?;
+    pub async fn new(path: &Path, blocks_per_file: NonZeroU64) -> Result<BlockStore> {
+        if fs::read_dir(path).await.is_err() {
+            fs::create_dir_all(path)
+                .await
+                .wrap_err("Failed to create Block Store directory.")?;
         }
         Ok(BlockStore {
             path: path.to_path_buf(),
+            blocks_per_file,
         })
     }
 
-    fn get_block_filename(block_height: u64) -> String {
-        format!("{}", block_height)
+    /// Returns expected filename of datafile containing block, provided its height
+    async fn get_block_filename(
+        dir: &Path,
+        block_height: NonZeroU64,
+        blocks_per_file: NonZeroU64,
+    ) -> Result<String> {
+        const INITIAL_DATA_FILE: &str = "1";
+        let storage_indices = storage_files_base_indices(dir)
+            .await
+            .wrap_err("Failed to list storage files")?;
+        let max_base = match storage_indices.iter().last() {
+            Some(max_base) => *max_base,
+            None => return Ok(INITIAL_DATA_FILE.to_owned()),
+        };
+        if block_height <= max_base {
+            return Err(eyre::eyre!("An attempt to random-access persistent storage (file except for the last one queued)."));
+        }
+        if (block_height.get() - max_base.get()) < blocks_per_file.get() {
+            return Ok(max_base.to_string());
+        }
+        Ok(block_height.to_string())
     }
 
     /// Get filysystem path for the block at height.
-    pub fn get_block_path(&self, block_height: u64) -> PathBuf {
-        self.path.join(BlockStore::get_block_filename(block_height))
+    ///
+    /// # Errors
+    /// - Filesystem access failure (HW or permissions)
+    ///
+    pub async fn get_block_path(&self, block_height: NonZeroU64) -> Result<PathBuf> {
+        let filename =
+            BlockStore::get_block_filename(&self.path, block_height, self.blocks_per_file)
+                .await
+                .wrap_err("Failed to fetch storage name for block")?;
+        Ok(self.path.join(filename))
     }
 
     /// Write block as file onto the disk.
     #[iroha_futures::telemetry_future]
     pub async fn write(&self, block: &VersionedCommittedBlock) -> Result<Hash> {
-        //filename is its height
-        let path = self.get_block_path(block.header().height);
-        let mut file = File::create(path)
+        let path = self
+            .get_block_path(
+                NonZeroU64::new(block.header().height)
+                    .wrap_err("An attempt to write a zero-height block.")?,
+            )
+            .await?;
+        let mut file = fs::OpenOptions::new()
+            .create(true)
+            .append(true)
+            .open(path)
```

Comment:
Could you wrap it with `bufwriter`? As for every write request it will make a future